index.js

```diff
@@ -0,0 +1,52 @@
+import fp from 'fastify-plugin'
+import onFinished from 'on-finished'
+import defaults from 'defaults'
+
+import { DEFAULT_OPTIONS, HEADERS } from './lib/constants.js'
+import { Store } from './lib/store.js'
+import { convertToMs } from './lib/helpers.js'
+
+const slowDownPlugin = async (fastify, settings) => {
+  const options = defaults(settings, DEFAULT_OPTIONS)
+  const store = new Store(convertToMs(options.timeWindow))
+
+  fastify.addHook('onClose', () => store.close())
+
+  fastify.addHook('onRequest', async (req, reply) => {
+    const value = store.incrementOnKey(options.keyGenerator(req))
+    const delay =
+      value > options.delayAfter
+        ? (value - options.delayAfter) * convertToMs(options.delay)
+        : 0
+
+    reply.header(HEADERS.limit, options.delayAfter)
+    reply.header(HEADERS.remaining, Math.max(options.delayAfter - value, 0))
+    if (delay !== 0) {
+      reply.header(HEADERS.delay, delay)
+      let timeout, resolve
+
+      const promise = new Promise(res => {
+        resolve = res
+        timeout = setTimeout(() => {
+          res()
+        }, delay)
```

Comment:
since we're only using the value we explicitly resolve this promise with, we can simplify this code even further

```suggestion
        timeout = setTimeout(res), delay)
```